### ISSUE TITLE
Introduce debug build of NativeScript

### ISSUE THREAD

**Issue by ivanbuhov** [NativeScript/ios-runtime#242]
We should have two  builds of NativeScript - for release and debug. The debug build will include asserts and checks at runtime to make debugging more painless. The project template should have the option to chose which version to use.

**Comment by jasssonpet** [comment]
@fealebenpae I have modified the build script in the [debug-build](https://github.com/NativeScript/ios-runtime/tree/jasssonpet/debug-build) branch to remove the debug symbols from the static library (which are several GBs big). Playing with Xcode 7 I don't seem to find the promised minification of the debug symbols of static libraries.

This doesn't impose an overhead on the current build process, because we are already building the runtime for debug with the static library on each commit.

How soon are we to drop iOS 7 support, because reading the "Debugging Info and Consistency" in [this article](http://landonf.org/code/ios/Radar_15800975_iOS_Frameworks.20140112.html) it would be better to have a debug build with the shared framework with dSYM files, which are small enough.

**Comment by fealebenpae** [comment]
For the sake of stack traces and everything, can we include the dSYM bundle for the release build of the framework?

**Comment by jasssonpet** [comment]
I want to ask if you just know, if we go the dSYM way for what architecture should they be or are there any "fat" dSYM bundles.

**Comment by fealebenpae** [comment]
Yes, the file in the `DWARF` folder inside a dSYM bundle is a fat Mach-O binary. We can use `lipo` to combine the dSYM bundles for the `iphoneos` and `iphonesimulator` platforms.

**Comment by jasssonpet** [comment]
Ok, I'll tweak our CMake config to generate dSYM bundles for the framework and merge all architectures in a single bundle.

**Comment by jasssonpet** [comment]
After generating dSYM bundles for the release build of the framework, I've found out that running `dsymutil NativeScript.framework/NativeScript -o NativeScript.dSYM` produced the (binary) same debug symbols, that were produced when building in `RelWithDebInfo` configuration.

I'll check if the archive step does anything special to strip them from the binary and leave them only in the bundle.

**Comment by jasssonpet** [comment]
The dSYM bundle for all architectures turns out to be 2.93GB.

**Comment by fealebenpae** [comment]
Okay, but is that for the Release build, or just RelWithDebInfo, or plain debug?

**Comment by jasssonpet** [comment]
This is for the `RelWithDebInfo`.

**Comment by fealebenpae** [comment]
Can we try to slim down the debug symbols by playing with Clang's `-g` flag?

**Comment by jasssonpet** [comment]
I'll have a look.

**Comment by jasssonpet** [comment]
After some further research here is what I found out:

* Xcode 6.4 - [`-g`](http://clang.llvm.org/docs/UsersManual.html#cmdoption-g) (Compiler default) - ~3GB for all architectures
* Xcode 7.0 - [`-g`](http://clang.llvm.org/docs/UsersManual.html#cmdoption-g) (Compiler default) - ~400MB for all architectures
* Xcode 7.0 - [`-gline-tables-only`](http://clang.llvm.org/docs/UsersManual.html#cmdoption-gline-tables-only) (Line tables only) - 80MB for all architectures

All dSYMs are nicely compressable (about 4x/5x times smaller).

**Comment by jasssonpet** [comment]
As I originally expected, there are some `ASSERT()` macros that rely on `#ifndef NDEBUG` fields:

* Field: [link](https://github.com/NativeScript/webkit/blob/6a4e3cbab6e5ef7d15bbc097426b6d29a45a3e61/Source/JavaScriptCore/heap/CopiedBlock.h#L162-L164)
* Assert: [link](https://github.com/NativeScript/webkit/blob/6a4e3cbab6e5ef7d15bbc097426b6d29a45a3e61/Source/JavaScriptCore/heap/CopiedBlock.h#L173)

**Comment by fealebenpae** [comment]
Alright, it seems that we will use the default Debug configuration. I propose that we limit debug symbols to `-gline-tables-only`, though.

**Comment by jasssonpet** [comment]
So in the end we will ship:

* NativeScript.framework + dSYM - built in `RelWithDebInfo` with `Line tables only` debug level, require Xcode 7
* libNativeScript.a built in `Release` without debug symbols
* libNativeScript.a built in `Debug` without debug symbols

Is that correct?

**Comment by fealebenpae** [comment]
Yes. Hopefully we will be able to ditch the static framework soon, or migrate to a distribution method that allows us to ship a larger package so as to include the debug symbols for the static library.

**Comment by jasssonpet** [comment]
The debug build of the static library is 170MB, and increases our package with 27MB when compressed.

**Comment by AntonioCuevaUrraco** [comment]
Will this issue make possible to see JS Errors on the crash reports?   
 or is it already possible somehow?   
I am trying to set up raygun https://github.com/pocketsmith/nativescript-raygun/issues/3

**Comment by ivanbuhov** [comment]
@AntonioCuevaUrraco Introducing debug build of NativeScript will not change the information in crash reports sent by Apple. For more info check https://github.com/NativeScript/ios-runtime/issues/554.

**Comment by mbektchiev** [comment]
Instead of separate debug build of NativeScript we intend to provide the `dSYM` package as an archive uploaded in the corresponding release tag in GitHub (e.g. https://github.com/NativeScript/ios-runtime/releases/v5.0.0)
